diff -ru admin-orig/runit-2.1.2/src/runit.c admin/runit-2.1.2/src/runit.c
--- admin-orig/src/runit.c	2014-08-10 15:22:35.000000000 -0300
+++ admin/src/runit.c	2019-06-30 23:21:45.947391610 -0300
@@ -299,6 +299,11 @@
   strerr_warn2(INFO, "sending KILL signal to all processes...", 0);
   kill(-1, SIGKILL);
 
+  strerr_warn2(INFO, "syncing...", 0);
+  //sync();
+  strerr_warn2(INFO, "exiting...", 0);
+  _exit(0);
+
   pid =fork();
   switch (pid) {
   case  0:
@@ -309,27 +314,27 @@
     reboot_system(RB_AUTOBOOT);
   }
   else {
-#ifdef RB_POWER_OFF
+#  ifdef RB_POWER_OFF
     strerr_warn2(INFO, "power off...", 0);
     sync();
     reboot_system(RB_POWER_OFF);
     sleep(2);
-#endif
-#ifdef RB_HALT_SYSTEM
+#  endif // RB_POWER_OFF
+#  ifdef RB_HALT_SYSTEM
     strerr_warn2(INFO, "system halt.", 0);
     sync();
     reboot_system(RB_HALT_SYSTEM);
-#else
-#ifdef RB_HALT
+#  else // RB_HALT_SYSTEM
+#    ifdef RB_HALT
     strerr_warn2(INFO, "system halt.", 0);
     sync();
     reboot_system(RB_HALT);
-#else
+#    else // RB_HALT
     strerr_warn2(INFO, "system reboot.", 0);
     sync();
     reboot_system(RB_AUTOBOOT);
-#endif
-#endif
+#    endif // RB_HALT
+#  endif // RB_HALT_SYSTEM
   }
   if (pid == 0) _exit(0);
   break;
@@ -337,7 +342,7 @@
   sig_unblock(sig_child);
   while (wait_pid(0, pid) == -1);
   }
-#endif
+#endif // RB_AUTOBOOT
 
   for (;;) sig_pause();
   /* not reached */
Only in admin/runit-2.1.2/src: .runit.c.swp
diff -ur runit-2.1.2-old/src/runit.c runit-2.1.2/src/runit.c
--- runit-2.1.2-old/src/runit.c	2019-12-02 18:36:02.982346238 -0400
+++ runit-2.1.2/src/runit.c	2019-12-02 18:46:24.483615039 -0400
@@ -66,6 +66,7 @@
   sig_block(sig_hangup);
   sig_block(sig_int);
   sig_catch(sig_int, sig_int_handler);
+  sig_catch(sig_term, sig_int_handler);
   sig_block(sig_pipe);
   sig_block(sig_term);
 
@@ -145,6 +146,7 @@
       sig_unblock(sig_child);
       sig_unblock(sig_cont);
       sig_unblock(sig_int);
+      sig_unblock(sig_term);
 #ifdef IOPAUSE_POLL
       poll(&x, 1, 14000);
 #else
@@ -156,6 +158,7 @@
       sig_block(sig_cont);
       sig_block(sig_child);
       sig_block(sig_int);
+      sig_block(sig_term);
       
       while (read(selfpipe[0], &ch, 1) == 1) {}
       while ((child =wait_nohang(&wstat)) > 0)
Only in runit-2.1.2-old/src: .runit.c.swp
